runtime/master/src/main/java/edu/snu/nemo/runtime/master/BlockManagerMaster.java

```diff
@@ -368,4 +357,82 @@ public void onMessageWithContext(final ControlMessage.Message message, final Mes
       }
     }
   }
+
+  /**
+   * The handler of block location requests.
+   */
+  @VisibleForTesting
+  public static final class BlockLocationRequestHandler {
+    private final String blockId;
+    private final CompletableFuture<String> locationFuture;
+
+    /**
+     * Constructor.
+     *
+     * @param blockId the ID of the block.
+     */
+    BlockLocationRequestHandler(final String blockId) {
+      this.blockId = blockId;
+      this.locationFuture = new CompletableFuture<>();
+    }
+
+    /**
+     * Completes the block location future.
+     * If there is any pending request, replies with the completed location.
+     *
+     * @param location the location of the block.
+     */
+    void complete(final String location) {
+      locationFuture.complete(location);
+    }
+
+    /**
+     * Completes the block location future with failure.
+     * If there is any pending request, replies with the cause.
+     *
+     * @param throwable the cause of failure.
+     */
+    void completeExceptionally(final Throwable throwable) {
+      locationFuture.completeExceptionally(throwable);
+    }
+
+    /**
+     * Registers a request for the block location.
+     * If the location is already known, reply the location instantly.
+     *
+     * @param requestId      the ID of the block location request.
+     * @param messageContext the message context to reply.
+     */
+    void registerRequest(final long requestId,
+                         final MessageContext messageContext) {
+      final ControlMessage.BlockLocationInfoMsg.Builder infoMsgBuilder =
+          ControlMessage.BlockLocationInfoMsg.newBuilder()
+              .setRequestId(requestId)
+              .setBlockId(blockId);
+
+      locationFuture.whenComplete((location, throwable) -> {
```

Comment:
There seems to be an assumption that the `locationFuture` is completed before this line is called. Please correct me if I'm wrong, but I don't see any multi-threaded operation with a `locationFuture` either. (in the sense that you use it to make an asynchronous request)

I'd remove `locationFuture` and `BlockLocationRequestHandler`, and instead use something else that is not associated with concurrent operations. It'd be nice if a simple `boolean`, or an `Optional` works for us.

Reply:
Sorry but I couldn't get your point. Why did you think that the future is always completed before the line? The future can be completed with block state changing, and it is independent with this method.

Reply:
Sorry I thought `messageContext.reply()` came after `locationFuture.whenComplete()`. I'll give it another look.